desc/objectives/utils.py

```diff
@@ -97,34 +97,52 @@ def factorize_linear_constraints(objective, constraint):  # noqa: C901
-        )
-        xp = put(xp, fixed_idx, b[fixed_rows])
-        # Some values might be fixed, but they still show up in other constraints
-        # this is where the fixed cols have >1 nonzero val.
-        # For fixed variables, we delete that row and col of A, but that means
-        # we need to subtract the fixed value from b so that the equation is balanced.
-        # e.g., 2 x1 + 3 x2 + 1 x3 = 4 ; 4 x1 = 2
-        # combining gives 3 x2 + 1 x3 = 3, with x1 now removed
-        b = put(
-            b,
-            unfixed_rows,
-            b[unfixed_rows] - A[unfixed_rows][:, fixed_idx] @ b[fixed_rows],
-        )
-    A = A[unfixed_rows][:, unfixed_idx]
-    b = b[unfixed_rows]
+    # will store the global index of the unfixed rows, idx
+    indices_row = np.arange(A.shape[0])
+    indices_idx = np.arange(A.shape[1])
+
+    # while loop has problems updating JAX arrays, convert them to numpy arrays
+    A = np.array(A)
+    b = np.array(b)
+    while len(np.where(np.count_nonzero(A, axis=1) == 1)[0]):
+        # fixed just means there is a single element in A, so A_ij*x_j = b_i
+        fixed_rows = np.where(np.count_nonzero(A, axis=1) == 1)[0]
+        # indices of x that are fixed = cols of A where rows have 1 nonzero val.
+        _, fixed_idx = np.where(A[fixed_rows])
+        unfixed_rows = np.setdiff1d(np.arange(A.shape[0]), fixed_rows)
+        unfixed_idx = np.setdiff1d(np.arange(A.shape[1]), fixed_idx)
+
+        # find the global index of the fixed variables of this iteration
+        global_fixed_idx = indices_idx[fixed_idx]
+        # find the global index of the unfixed variables by removing the fixed variables
+        # from the indices arrays.
+        indices_idx = np.delete(indices_idx, fixed_idx)  # fixed indices are removed
+        indices_row = np.delete(indices_row, fixed_rows)  # fixed rows are removed
+
+        if len(fixed_rows):
+            # something like 0.5 x1 = 2 is the same as x1 = 4
+            b = put(b, fixed_rows, b[fixed_rows] / np.sum(A[fixed_rows], axis=1))
+            A = put(
+                A,
+                Index[fixed_rows, :],
+                A[fixed_rows] / np.sum(A[fixed_rows], axis=1)[:, None],
+            )
+            xp = put(xp, global_fixed_idx, b[fixed_rows])
+            # Some values might be fixed, but they still show up in other constraints
+            # this is where the fixed cols have >1 nonzero val.
+            # For fixed variables, we delete that row and col of A, but that means
+            # we need to subtract the fixed value from b so that the equation is
+            # balanced.
+            # e.g., 2 x1 + 3 x2 + 1 x3 = 4 ; 4 x1 = 2
+            # combining gives 3 x2 + 1 x3 = 3, with x1 now removed
+            b = put(
+                b,
+                unfixed_rows,
+                b[unfixed_rows] - A[unfixed_rows][:, fixed_idx] @ b[fixed_rows],
+            )
+        A = A[unfixed_rows][:, unfixed_idx]
```

Comment:
ie, suppose row 3 is fixed, and row 5 will be fixed once row 3 is removed. On the first iteration, row 3 is removed, but that means what was row 5 is now row 4?

Reply:
Yes, I handle that change with the `global_fixed_idx` for `xp` inside the loop, since that is the only array that doesn't change its size. And after the loop, I use `indices_idx` which holds the global indices of the whole unfixed parameters.

Reply:
I'm not sure that fixes the issue. `global_fixed_idx = indices_idx[fixed_idx]`, but `fixed_idx` depends on `A` which is changing size?